### PR TITLE
feat(components): add tree select component

### LINKED ISSUE #2037: [Feature Request] tree-select
### Existing Component
No

### Component Name
tree-select

### Description
建议增加tree-select组件

### LINKED ISSUE #6887: [Feature Request] [New Component] [] 希望增加下拉树
Feature type: **`New Component`**

## Details

- New component name: `下拉树`

## Description

希望增加下拉树

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by yujinpan**
The tree selector of the drop-down menu, it combines the functions of components `el-tree` and `el-select`.

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by JeremyWuuuuu** [comment]
@yujinpan Run 

```shell
$ pnpm lint
```

for fixing the linter errors

**Comment by yujinpan** [comment]
@JeremyWuuuuu sorry, I fixed it.

File: packages/components/tree-select/src/select.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  {
+    tree,
+    key,
+  }: {
+    select: Ref<InstanceType<typeof ElSelect> | undefined>
+    tree: Ref<InstanceType<typeof ElTree> | undefined>
+    key: Ref<string>
+  }
+) => {
+  const result = {
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/select.ts:19]:
```ts
import { useNamespace } from '@element-plus/hooks'

const ns = useNamespace('tree-select')

const classes = ns.e('popper')

see `packages/hooks/use-namespace/index.ts` file for more detail.
```

**Review by jw-foss — COMMENTED** [review]
I left a few comments for you first. ;) 🏓  be back later.

File: docs/en-US/component/tree-select.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
+---
+title: TreeSelect
+lang: en-US
+---
+
+# TreeSelect
+
+The tree selector of the drop-down menu,
```
[the remark this answers is not in the document]
Comment (yujinpan) [reply docs/en-US/component/tree-select.md]: https://github.com/yujinpan/element-plus/commit/f73bdfcccc4d4c1132ac8a36535317b8e34103de

File: packages/components/tree-select/src/select.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { computed, nextTick, toRefs } from 'vue'
+import { pick } from 'lodash-unified'
+import ElSelect from '../../select/src/select.vue'
```
[the remark this answers is not in the document]
Comment (yujinpan) [reply packages/components/tree-select/src/select.ts]: https://github.com/yujinpan/element-plus/commit/fa8e6c2204ad6f874a6bd5d62d3e47ac85eca52b

File: packages/components/tree-select/src/select.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import type ElTree from '../../tree/src/tree.vue'
```
[the remark this answers is not in the document]
Comment (yujinpan) [reply packages/components/tree-select/src/select.ts]: https://github.com/yujinpan/element-plus/commit/fa8e6c2204ad6f874a6bd5d62d3e47ac85eca52b

File: packages/components/tree-select/src/select.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import type { Ref } from 'vue'
```
[the remark this answers is not in the document]
Comment (yujinpan) [reply packages/components/tree-select/src/select.ts:5]: https://github.com/yujinpan/element-plus/commit/fa8e6c2204ad6f874a6bd5d62d3e47ac85eca52b

File: packages/components/tree-select/src/tree-select-option.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { defineComponent } from 'vue'
+import ElSelectOption from '@element-plus/components/select/src/option.vue'
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/tree-select-option.ts:2]:
```suggestion
import { ElOption } from '@element-plus/components/select'
```

File: packages/components/tree-select/src/tree-select-option.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+const component = defineComponent({
+  extends: ElSelectOption,
+  setup(props, ctx) {
+    const result = (ElSelectOption.setup as NonNullable<any>)(props, ctx)
+
+    delete result.selectOptionClick
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/tree-select-option.ts:9]:
```suggestion
    result.selectOptionClick = null // or undefined.
```
Would this seem better to you?

File: packages/components/tree-select/src/tree-select.vue [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    const select = ref<InstanceType<typeof ElSelect>>()
+    const tree = ref<InstanceType<typeof ElTree>>()
+
+    const key = computed(() => props.valueKey || props.nodeKey || 'value')
+
+    const selectProps = useSelect(props, context, { select, tree, key })
+    const treeProps = useTree(props, context, { select, tree, key })
+
+    // expose ElTree/ElSelect methods
+    const methods = reactive({})
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/tree-select.vue:27]: **Praise:** this is clever!

File: packages/components/tree-select/__tests__/tree-select.spec.ts [hunk 9]
```diff
@@ -0,0 +1,312 @@
+import { nextTick, ref, h } from 'vue'
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/__tests__/tree-select.spec.ts:1]:
```suggestion
import { h, nextTick, ref } from 'vue'
```

File: packages/components/tree-select/src/tree-select.vue [hunk 10]
```diff
@@ -0,0 +1,76 @@
+<script lang="ts">
+import { defineComponent, ref, h, reactive, computed, onMounted } from 'vue'
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/tree-select.vue:2]:
```suggestion
import { computed, defineComponent, h,  onMounted, ref, reactive } from 'vue'
```

File: packages/components/tree-select/src/tree.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    value: key.value,
+    ...props.props,
+  }))
+
+  const getNodeValByProp = (
+    prop: 'value' | 'label' | 'children' | 'disabled' | 'isLeaf',
+    data: TreeNodeData
+  ) => {
+    const propVal = propsMap.value[prop]
+    if (propVal instanceof Function) {
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/tree.ts:58]:
```suggestion
import { isFunction } from '@element-plus/utils'
    if (isFunction(propVal)) {
```

File: packages/components/tree-select/src/tree-select.vue [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+      })
+    })
+
+    return () =>
+      h(
+        ElSelect,
+        reactive({
+          ...selectProps,
+          ref: (ref) => (select.value = ref),
+        }),
```
Comment (JeremyWuuuuu) [inline packages/components/tree-select/src/tree-select.vue:61]: **Question:** was this really necessary? I'm not sure, would you elaborate a little bit about this?

**Review by jw-foss — COMMENTED** [review]
@yujinpan Left a few comments for you, in general the code LGTM, thanks! 🏓

File: docs/examples/tree-select/filterable.vue [hunk 13]
```diff
@@ -0,0 +1,104 @@
+<template>
+  <el-tree-select v-model="value" :data="data" filterable />
+  <el-divider></el-divider>
```
Comment (JeremyWuuuuu) [inline docs/examples/tree-select/filterable.vue:3]:
```suggestion
  <el-divider />
```

File: docs/examples/tree-select/filterable.vue [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  filter method:
+  <el-tree-select
+    v-model="value"
+    :data="data"
+    :filter-method="filterMethod"
+    filterable
+  />
+  <el-divider></el-divider>
```
Comment (JeremyWuuuuu) [inline docs/examples/tree-select/filterable.vue:11]:
```suggestion
  <el-divider />
```

File: docs/examples/tree-select/multiple.vue [hunk 15]
```diff
@@ -0,0 +1,84 @@
+<template>
+  <el-tree-select v-model="value" :data="data" multiple />
+  <el-divider></el-divider>
```
Comment (JeremyWuuuuu) [inline docs/examples/tree-select/multiple.vue:3]:
```suggestion
  <el-divider />
```

File: docs/examples/tree-select/slots.vue [hunk 16]
```diff
@@ -0,0 +1,104 @@
+<template>
+  <el-tree-select v-model="value" :data="data">
+    <template #default="{ data: { label } }">
+      {{ label }}<span style="color: gray">(suffix)</span></template
+    >
+  </el-tree-select>
+  <el-divider></el-divider>
```
Comment (JeremyWuuuuu) [inline docs/examples/tree-select/slots.vue:7]:
```suggestion
  <el-divider />
```

**Review by jw-foss — COMMENTED** [review]
Additional comments

File: docs/.vitepress/crowdin/en-US/pages/component.json [hunk 17]
```diff
@@ -206,6 +206,10 @@
         "link": "/tree",
         "text": "Tree"
       },
+      {
+        "link": "/tree-select",
+        "text": "TreeSelect"
+      },
```
Comment (JeremyWuuuuu) [inline docs/.vitepress/crowdin/en-US/pages/component.json:212]:
```suggestion
      {
        "link": "/tree-select",
        "text": "TreeSelect",
        "promotion": "2.1.8"
      },
```

File: packages/components/tree-select/src/tree-select-option.ts (hunk 6 above)
Comment (yujinpan) [reply packages/components/tree-select/src/tree-select-option.ts]: https://github.com/yujinpan/element-plus/commit/660ed13987dadefb18994cdb5e71b14594e27be3

File: packages/components/tree-select/src/tree-select-option.ts (hunk 7 above)
Comment (yujinpan) [reply packages/components/tree-select/src/tree-select-option.ts:10]: - The `setup` result will overrides `methods` when the property exists.
- Defined in `methods` for better access to `this`.

File: packages/components/tree-select/src/tree-select.vue (hunk 8 above)
Comment (yujinpan) [reply packages/components/tree-select/src/tree-select.vue:27]: 😄

File: packages/components/tree-select/__tests__/tree-select.spec.ts (hunk 9 above)
Comment (yujinpan) [reply packages/components/tree-select/__tests__/tree-select.spec.ts]: https://github.com/yujinpan/element-plus/commit/c019adfb206c4d5414662b9978bb79451a97832e

File: packages/components/tree-select/src/tree-select.vue (hunk 10 above)
Comment (yujinpan) [reply packages/components/tree-select/src/tree-select.vue]: https://github.com/yujinpan/element-plus/commit/c019adfb206c4d5414662b9978bb79451a97832e

File: packages/components/tree-select/src/tree.ts (hunk 11 above)
Comment (yujinpan) [reply packages/components/tree-select/src/tree.ts]: https://github.com/yujinpan/element-plus/commit/a7ab3d81b64dec2893fca6024b7412ea8d2eb0c2

File: packages/components/tree-select/src/tree-select.vue (hunk 12 above)
Comment (yujinpan) [reply packages/components/tree-select/src/tree-select.vue:68]: https://github.com/yujinpan/element-plus/commit/6d148dad3a11f89c3731a92124c3c9f41ae58276

1. The `props` is processed into `Refs`, but `v-bind` and render function props cannot read `Refs`, so use `reactive` unwrap the `Refs` and keep reactive.
2. The keyword `ref` requires `Ref`, but `reactive` broke it, so use function.

File: docs/examples/tree-select/filterable.vue (hunk 13 above)
Comment (yujinpan) [reply docs/examples/tree-select/filterable.vue]: https://github.com/yujinpan/element-plus/commit/f6d59a60eaf21d7250c7f78ea10e35894888ac5d

File: docs/examples/tree-select/filterable.vue (hunk 14 above)
Comment (yujinpan) [reply docs/examples/tree-select/filterable.vue]: https://github.com/yujinpan/element-plus/commit/f6d59a60eaf21d7250c7f78ea10e35894888ac5d

File: docs/examples/tree-select/multiple.vue (hunk 15 above)
Comment (yujinpan) [reply docs/examples/tree-select/multiple.vue]: https://github.com/yujinpan/element-plus/commit/f6d59a60eaf21d7250c7f78ea10e35894888ac5d

File: docs/examples/tree-select/slots.vue (hunk 16 above)
Comment (yujinpan) [reply docs/examples/tree-select/slots.vue]: https://github.com/yujinpan/element-plus/commit/f6d59a60eaf21d7250c7f78ea10e35894888ac5d

File: docs/.vitepress/crowdin/en-US/pages/component.json (hunk 17 above)
Comment (yujinpan) [reply docs/.vitepress/crowdin/en-US/pages/component.json]: https://github.com/yujinpan/element-plus/commit/92cf3853bd0db1d6b8d0e379f09918cb97a5f2c6

**Review by jw-foss — APPROVED** [review]
Good job. Approved.

**Pull request merged**